### PR TITLE
tests/smoke: conditional per-test config isolation for the UI tier

### LINKED ISSUE #810: smoke/ui: per-test config reset for the UI tier (deferred #576 keystone STEP 5, #582 section C)
Follow-up to #582 (section C — the deferred keystone STEP 5; per-test fixes landed in #799).

The UI tier (`tests/smoke/ui/`) shares a session-scoped login/box/cookies; only `browser_page` is per-test. A `ui_e2e`/`ui_browser` POST that mutates config persists into every later UI test. The module-scoped baseline fixture (`_pfb_module_baseline`, tests/smoke/conftest.py) explicitly excludes the `ui` subpackage with a docstring claiming the UI tier is "reset separately" — that separate mechanism was never built.

Scope:

- Add a marker-gated, function-scoped autouse finalizer in `tests/smoke/ui/conftest.py` calling `reset_pfb_baseline(deployed_vm)` after each mutating test (`ui_e2e`/`ui_browser` only; `ui_render` stays read-only on shared state). Preserve the injected PHPSESSID cookie; clear `localStorage`/`sessionStorage`.
- `reset_pfb_baseline` costs a `php_eval` (section deletes + `write_config` + `services_unbound_configure`) plus conditional CLI round-trips per call — seconds each across 60+ mutating tests. Measure the added wall clock with a dispatch run and size the gating (e.g. skip after tests that provably self-restore) before landing.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

### LINKED ISSUE #964: smoke/ui: test_gateway_dnsbl_lenient_save_roundtrip stale since a2c5c26a (writeSection adapter coalesce stores 'off', test expects '')
`tests/smoke/ui/test_browser_dnsbl.py::test_gateway_dnsbl_lenient_save_roundtrip` asserts that saving the DNSBL page with the `pfb_dnsbl_lenient` checkbox unchecked stores `''` (its docstring: "the only page-reachable stored tokens are 'on' (checked) and '' (unchecked)").

That was true when the test landed, but a2c5c26a ("pfblockerng: apply per-field write adapters in PfbConfig::writeSection", PR #949) landed the same day and changed the semantics: the DNSBL page persists via `PfbConfig::writeSection` (pfblockerng_dnsbl.php:903), which now round-trips every registered key through its read+write adapters — and `PfbLenient` coalesces `''` to the canonical `'off'` (verified executable: `pfb_cfg_lenient_write(pfb_cfg_lenient_read(''))` returns `'off'`). So an unchecked save now deterministically stores `'off'`, and the test's restore assertion `stored == ''` can never pass.

Evidence:

- Nightly ui-tests run 28851771590 (devel `30914cd`, before a2c5c26a): test passed — `writeSection` at that commit was a plain `config_set_path`, no adapter ride.
- Runs 28900064099 / 28907688229 (branch based on devel `60fb4edb`, which includes a2c5c26a): both browser legs fail with `AssertionError: pfb_dnsbl_lenient gateway FAIL after restore: stored 'off', expected ''`. The branch's diff touches no `src/` file, so the regression is the base's.

The browser tier is dispatch-only, so devel's own nightly will show the same failure on its next run.

Scope check of the class (every registered key with both adapters × pages saving via `writeSection`): `PfbToggle`'s off token is `''` (identity — unaffected); `pfb_idn` is already written canonically by the page; `pfb_keep`'s roundtrip test already expects `'off'`; the `ui_e2e` toggle flows normalize with `or flow.off`. Only this one test asserts the stale `''` vocabulary.

Fix: update the test's expected stored token for the unchecked state to the canonical `'off'` (and its docstring NOTE) — the DOM oracle half is unaffected (`pfb_cfg_toggle_read('off')` renders unchecked). Fix rides the issue #810 branch (`claude/gh-issue-810-0xngul`), whose mandated verification dispatch is blocked by this failure.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

### PR DESCRIPTION
**Pull request by andrebrait**
Implements the deferred #576 keystone STEP 5 (#582 section C): per-test config isolation for the ADR-14 UI tier, which shares one session-scoped login+box across ~170 mutating `ui_e2e`/`ui_browser` tests with no reset mechanism — `_pfb_module_baseline` excludes `tests/smoke/ui/` and pointed at a per-test mechanism that was never built.

## Design (sized for the performance concern on the issue)

An unconditional reset after every mutating test is unaffordable, so the autouse fixture `_ui_pfb_isolation` probes a **one-ssh config digest** at teardown (`helpers.pfb_config_digest`: md5 of `config.xml` with the volatile `<revision>` block stripped — false-dirty is safe, false-clean is not) and acts **only when a test actually left config dirty**:

- **Clean path** (most tests — the drive-form-back/finally-restore idiom): one `cat /conf/config.xml` over ssh, **~0.1s measured**.
- **Dirty path**: `helpers.restore_pfb_config_baseline` — copy back a guest-side `config.xml` snapshot taken at session baseline + drop `/tmp/config.cache` (the exact idiom upstream `config.lib.inc` uses on a config replace), then `services_unbound_configure` + `clearip`/`cleardnsbl` (package-presence gated) + a blocking `apply_filter_sync`. **~2.5s measured**, paid only by tests that genuinely leak (~40 in the functional leg).

Restore-to-session-baseline, **not** `reset_pfb_baseline`'s wipe-to-empty: the wipe fits the module-scoped smoke tier (each module re-establishes its infra) but under the UI tier's single session it deleted `pfb_dnsvip4`/ports with nothing left to re-add them, so every later `pfblockerng_dnsbl.php`/SafeSearch save failed `pfb_validate_vips` (run 28900064099, 23 functional + 1 browser failures — all reading back `''`). `deployed_vm` now establishes the DNSBL VIP once (idempotent), so it is part of the tier baseline; `browser_page` clears `localStorage`/`sessionStorage` per test (PHPSESSID untouched — it lives on the session-scoped `browser_context`).

Second commit fixes #964, found by this work: a2c5c26a routed section saves through the PfbConfig write adapters, so an unchecked `pfb_dnsbl_lenient` save stores the canonical `'off'` (verified: `pfb_cfg_lenient_write(pfb_cfg_lenient_read(''))` → `'off'`); the gateway roundtrip test still asserted the pre-adapter `''` vocabulary and failed on every browser dispatch since.

## Verification (live-VM dispatches, scope=full, CE + Plus)

Runs: 28907688229 (all tiers, isolation commit) + 28913116113 (browser tier, after the #964 test fix) — **all legs green**. Off-box: 14 new unit tests (digest semantics incl. hostile rows: no/second `<revision>` block, CRLF, empty, ssh-failure raise paths; restore sequence/guards), red→green executed both directions.

Wall clock, pytest step vs the last full-scope devel nightly (28851771590):

| leg | render | functional (`ui_e2e`) | browser |
|---|---|---|---|
| ce-2.8 | 69s → 70s | 318s → 620s | 132s → 148s |
| plus-26.03 | 66s → 68s | 389s → 706s | 128s → 157s |

The functional legs pay ~+5 min each — ~40 non-self-restoring tests × restore + 128 probes, plus the resolver reconfigures those restores trigger. These suites are dispatch-only (not PR-gating); a follow-up can shrink this by making the leaky tests self-restore.

Fixes #810.
Fixes #964.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Fh8xjHdECzVmC9Ed5TmD9p

---
_Generated by [Claude Code](https://claude.ai/code/session_01Fh8xjHdECzVmC9Ed5TmD9p)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved UI smoke isolation by detecting config drift and automatically restoring the UI baseline when tests mutate `/conf/config.xml`.
  * Hardened browser test cleanup by clearing per-origin storage to prevent cross-test leakage.

* **Tests**
  * Added dedicated unit coverage for UI config digesting and revision stripping, including multiple edge cases.
  * Extended restore/baseline sequencing tests and updated DNSBL UI behavior assertions to match canonical saved tokens.
  * Re-tagged affected UI rendering tests to run under the stricter isolation probe.

### CODE REVIEW

File: tests/smoke/helpers.py [hunk 1]
```diff
@@ -2739,6 +2740,114 @@ def reset_pfb_baseline(vm: SmokeVM, *, timeout: float = 300.0) -> None:
...
+       config.
+
+    Issue #810.
+    """
+    cp_result = vm.ssh(
+        "/bin/sh",
+        "-c",
+        f"cp {snapshot_path} /conf/config.xml && rm -f /tmp/config.cache",
+        timeout=timeout,
+    )
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/smoke/helpers.py:2788]: Applied in ad588f3b, with one adjustment to the suggested fix: passing the argv directly wouldn't work here — `SmokeVM.ssh`'s multi-arg convention `shlex.join`s the tokens, so the `&&` would reach `cp` as a literal argument. The correct house form is the documented **single-string** convention (one argument runs verbatim under the guest `/bin/sh`), which drops the redundant `"/bin/sh", "-c"` layer, plus `shlex.quote(snapshot_path)` for the caller-supplied path. Verified: the restore unit rows (k)-(n) still pass unchanged (they assert on the joined command substring).

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/smoke/ui/conftest.py [hunk 2]
```diff
@@ -285,6 +296,110 @@ def webui(deployed_vm: SmokeVM, admin_credentials: tuple[str, str], webgui_proto
...
+    established the DNSBL VIP, so the baseline already includes both. The guest-side
+    copy to :data:`UI_CONFIG_SNAPSHOT` IS the restore artifact
+    :func:`~tests.smoke.helpers.restore_pfb_config_baseline` copies back on the dirty
+    path -- taken back-to-back with the digest, no writer running in between, so the
+    two describe the exact same config state. Mutable (a dict, not a bare str) so
+    :func:`_ui_pfb_isolation` can re-snapshot the digest in place after a restore.
+    """
+    from .. import helpers
+
+    result = smoke_vm.ssh("/bin/sh", "-c", f"cp /conf/config.xml {UI_CONFIG_SNAPSHOT}")
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/smoke/ui/conftest.py]: Applied in ad588f3b using the single-string `SmokeVM.ssh` convention (one argument runs verbatim under the guest `/bin/sh`) — both paths here are fixed internal constants, so no quoting is needed, and the redundant `"/bin/sh", "-c"` layer is gone.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by andrebrait** [comment]
## Review audit trail (pr-merge-flow)

**Review shape:** `review-fanout` multi-agent workflow (3 lenses → dedup → execution-grounded adversarial verify per finding, agents at effort `xhigh`), chosen because the PR exceeds the large-PR threshold (635 changed lines across 5 files). 3 findings confirmed, 0 refuted, per-file verdicts complete (5 files × 3 lenses).

**Reviewer status:** CodeRabbit — rate-limited at PR open ("next review in 11 minutes" > the 5-minute rule) → dropped from this flow; will be folded in if it reviews late. GitHub Copilot — requested, reviewed (2 inline comments). Snyk — no status on the head SHA (not active on this PR).

### Findings ledger

| # | Source | Severity | Finding | Outcome |
|---|--------|----------|---------|---------|
| 1 | fanout | nitpick | Docstrings cited run 28900064099 as "23 functional + 1 browser, all reading back `''`" — the browser failure was actually the pre-existing #964 bug (`'off'`), not the VIP-wipe class | **fixed@1d7528f9** (reworded both docstrings) |
| 2 | fanout | **blocking** | Two `ui_render`-marked tests in `test_render_smoke.py` (`test_ip_page_never_leaks_maxmind_key`, `test_alerts_unified_log_colour_fields_render`) mutate config.xml as setup and escaped the marker-gated isolation net; a failed finally-restore would leak undetected — exactly the #810 failure class | **fixed@1d7528f9** (dual-marked `ui_e2e`; they stay in the Tier-A gate via the module's `ui_render` marker and now ride the probe) |
| 3 | fanout | nitpick | Inverse marker mix: two read-only `test_category_edit.py` tests dual-marked `ui_e2e`(module)+`ui_render`(function) pay a probe, contradicting the docstring's "render = no probe cost" | **code SKIPPED** (probe-on-dual-mark is the fail-safe direction; cost is 2 × ~0.1s — precise own-marker logic isn't worth it) / **docstring fixed@1d7528f9** (gating rule now stated exactly) |
| 4 | Copilot | comment | `restore_pfb_config_baseline` shells `/bin/sh -c` through `SmokeVM.ssh` (which already wraps in `/bin/sh -c`) with an unquoted caller-supplied path | **fixed@ad588f3b** (single-string ssh convention + `shlex.quote(snapshot_path)`; the argv-direct suggestion was adjusted — `shlex.join` would make the `&&` a literal `cp` argument) |
| 5 | Copilot | comment | Same redundant `"/bin/sh", "-c"` layer in `_ui_config_baseline`'s snapshot `cp` | **fixed@ad588f3b** (single-string convention; both paths are fixed constants) |

Fix-delta re-review: the two review-fix commits are 2 marker decorators + docstring rewording + an ssh-call-shape change with unchanged unit assertions (15/15 pass) — below the non-trivial threshold; no focused re-review spawned. Final verification: `ui-tests.yml` dispatch (scope=full, tier=all, CE+Plus) re-queued on head ad588f3b (the run on the intermediate commit was cancelled); merge proceeds only on green.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by andrebrait** [comment]
Ledger update (pre-merge catch-all sweep): CodeRabbit's late review (on 1d7528f9) landed one nitpick — `usedforsecurity=False` on the digest md5 — **fixed@3cf43de9** (unit rows still 15/15; the flag changes no digest value, it documents the ADR-42 self-comparison intent and silences the CWE-327 SAST false positive). Its rate-limit notice for the ad588f3b..3cf43de9 delta states "next review available in 53 minutes" (> the 5-minute rule), so it is not waited on further; that delta is a review-fix kwarg/ssh-shape change already covered by Copilot's review and the unchanged unit suite. All five earlier ledger items remain as recorded. No live-VM re-dispatch for 3cf43de9: the commit is a runner-side kwarg with identical digest output, pinned by the off-box unit rows; the live verification stands on run 28914748394 (all legs green, scope=full, CE+Plus).

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_
